### PR TITLE
Polo refactoring (+ fix rounding)

### PR DESCRIPTION
**Pull request by Ofekw**
ExchangePoloniexAPI is getting too complicated to edit safely. Begin refactoring to bring it under test.

Pulled out MakeRequest into a RequestHelper to make it easier to mock during unit testing. The helper is now initialized during the api constructor or passed in.

Also refactored parse order details in Polo. Lots of duplicate code removed. Added code coverage for everything that was refactored.

### CODE REVIEW

File: ExchangeSharp/API/IRequestHelper.cs [hunk 1]
```diff
@@ -0,0 +1,18 @@
+﻿namespace ExchangeSharp.API.Services
```
Comment (bichuga) [inline ExchangeSharp/API/IRequestHelper.cs]: Change namespace to remove .Services

File: ExchangeSharp/API/APIException.cs [hunk 2]
```diff
@@ -0,0 +1,23 @@
+﻿namespace ExchangeSharp
```
Comment (bichuga) [inline ExchangeSharp/API/APIException.cs]: don't forget header

File: ExchangeSharp/API/Exchanges/ExchangeBinanceAPI.cs [hunk 3]
```diff
@@ -22,6 +22,8 @@ MIT LICENSE
 
 namespace ExchangeSharp
 {
+    using ExchangeSharp.API.Services;
```
Comment (bichuga) [inline ExchangeSharp/API/Exchanges/ExchangeBinanceAPI.cs]: Remove all of these excess usings after namespace is fixed

File: ExchangeSharp/API/Exchanges/ExchangePoloniexAPI.cs [hunk 4]
```diff
@@ -302,13 +308,16 @@ public override IEnumerable<ExchangeMarket> GetSymbolsMetadata()
...
+        }, ...
              */
 
             var markets = new List<ExchangeMarket>();
             Dictionary<string, JToken> lookup = MakeJsonRequest<Dictionary<string, JToken>>("/public?command=returnOrderBook&currencyPair=all&depth=0");
             foreach (var kvp in lookup)
             {
+                // StepSize is 8 decimal places for both price and amount on everything at Polo
+                const decimal StepSize = 0.00000001m;
+                const decimal minTradeSize = 0.0001m;
```
Comment (bichuga) [inline ExchangeSharp/API/Exchanges/ExchangePoloniexAPI.cs]: Move outside the foreach

File: ExchangeSharp/API/RequestHelper.cs [hunk 5]
```diff
@@ -0,0 +1,82 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace ExchangeSharp.API.Services
+{
+    using System.IO;
+    using System.Net;
+
+    class RequestHelper : IRequestHelper
```
Comment (bichuga) [inline ExchangeSharp/API/RequestHelper.cs]: public

Comment (bichuga) [inline ExchangeSharp/API/RequestHelper.cs]: Add documentation

**Comment by jjxtra** [comment]
Can we make the RequestHelper a public get; set; property and have it default to the http implementation? If set to null it should choose the default http implementation as well.

This way constructor overloads are not necessary, and a mocking framework can freely change it.

I'd also like to put any dependancy methods that the request helper is using behind an interface instead, that way they aren't publicly exposed to anyone using an instance of an API class. Having them protected served this purpose, but now they are dependant methods of another class, we need an interface. We will need to define an IRequestHelper interface, or some better name if you think of one :)

Example: Uri IRequestHelper.ProcessRequestUrl(UriBuilder url, Dictionary<string, object> payload);

The method only shows up if the instance is of type IRequestHelper.

Let me know your thoughts, counter-arguments, changes, etc.

**Comment by bichuga** [comment]
@jjxtra You have it right that all of the request helper stuff needs to move out of BaseAPI. The plan is to put them in IRequestHelper but once we do we need to also move each API's ProcessReq and ProcessResp logic. These would probably be implemented as subclasses of RequestHelper. This refactor is already enormous just to get Polo testable so we decided to cap it here. Get this in so people can start unit testing their contributions.

I don't have a strong opinion of using a property or a constructor parameter for setting the IRequestHelper.  (I'm not even internally consistent on this). Would be happy to switch that if you prefer.

**Comment by Ofekw** [comment]
@jjxtra I'll have a think about what you said and have a code update on here tomorrow no worries. Cheers for the feedback.

**Comment by jjxtra** [comment]
Let's switch to the property. We can leave the methods public and do a follow up pull request.

**Comment by jjxtra** [comment]
Merging this pull request. I'll do a follow up checkin.

**Comment by jjxtra** [comment]
I've pushed my changes. Request helper is now APIRequestMaker and I also added IAPIRequestHandler interface to only give the APIRequestMaker class the minimum it needs to make a request. Let me know any thoughts or ideas for improvement.

**Comment by jjxtra** [comment]
I think we are good on this refactor for now. I like the protected methods with override. May revisit layer but unit tests are easy now.

**Pull request merged**